### ISSUE TITLE
[performance] High lag on long lines

### ISSUE THREAD

**Issue by Aerijo** [atom/autocomplete-plus#949]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

When typing on a very long line (e.g., 500+ letter characters) there is noticeable input lag. I say letter, because a majority of space characters seem to increase this threshold. However, the performace lag only occurs when a space is one of the characters near the end of the line (lag occurs when typing past this space).

For me, this is caused by the function [`getLegacyPrefix`](https://github.com/atom/autocomplete-plus/blob/5b7f35384c570aed801d0412a0e2684ec45f50c7/lib/autocomplete-manager.js#L519-L526), but I believe the issue would persist even if _all_ providers were to update to version 4. 

Specifically, executing the prefix search regex on the entire line is causing the lag. My suggestion would be to, before executing the search, check the line length and see if it's above some threshold length (100 would be good). If it is, only search against the last number of characters equal to the threshold length.

For example,
```js
getLegacyPrefix (editor, bufferPosition) {
    let line = editor.getTextInRange([[bufferPosition.row, 0], bufferPosition]) // change to let to allow reassignment
    let threshold = 100 // this can be made an object property and the package setting observed
    if (line.length > threshold) { // check length
      line = line.split(-threshold) // trim to max length if needed
    }
    const prefix = this.prefixRegex.exec(line) // much faster, regardless of actual line length
    if (!prefix || !prefix[2]) {
      return ''
    }
    return prefix[2]
  }
```
I tested this change, and it appears to work fine. It should not affect any current providers. If some did rely on such long prefixes, they should also be experiencing the lag that comes with it. To be safe though, the threshold could
- be made configurable in this package's settings, or
- be made configurable for this package, but set by a provider (i.e., this package uses the maximum value of all registered providers for the language), or
- not be made configurable. Instead, if some provider requires an extremely long prefix, they can get the prefix themselves, much like providers requiring custom prefixes already do.

I prefer the first option.

### Steps to Reproduce

1. Open Atom in safe mode
2. Copy and paste the following
```
abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghi type here -> 
```
3. Observe lag when typing many letters in quick succession.

**Expected behaviour:** [What you expect to happen]
No input lag, as happens when this package is disabled

**Actual behaviour:** [What actually happens]
Input lag, which the profile at the bottom shows is caused by the `this.prefixRegex.exec(line)` function in `getLegacyPrefix()`

**Reproduces how often:** [What percentage of the time does it reproduce?]
100%

Some times it seems to work fine when it was lagging just moments ago. However, I was able to consistently reproduce the behaviour by following the instructions I give, even after a computer restart.

### Versions
```
Atom    : 1.23.3
Electron: 1.6.15
Chrome  : 56.0.2924.87
Node    : 7.4.0

apm  1.18.12
npm  3.10.10
node 6.9.5 x64
atom 1.23.3
python 2.7.10
git 2.14.3

macOS 10.13.3
```

### Additional Information
Profile taken that shows the extreme lag. Extension needs to be changed to `.cpuprofile` (I couldn't find a way to provide one directly).

[profile.txt](https://github.com/atom/autocomplete-plus/files/1670833/profile.txt)

**Comment by ungb** [comment]
👋  @Aerijo, You shouldn't be able to have over 500 characters on 1 line by default.  Did you change the setting for this within `settings-> editor`?

**Comment by Aerijo** [comment]
@ungb buffer lines, not screen lines. Buffer lines can be arbitrarily long (eg. think of a minified file)

**Comment by ungb** [comment]
Hey @Aerijo,

This is what I see when I try your repro steps.   With even a minified line it should soft warp by default, if it doesn't it can be that you increase the soft wrap setting.  

![longline](https://user-images.githubusercontent.com/5624870/36182405-5b4c8fc2-10de-11e8-92b2-3e85eeba5346.gif)


Any chance you can check the settings from https://github.com/atom/autocomplete-plus/issues/949#issuecomment-363923684?

**Comment by Aerijo** [comment]
@ungb Yes, the screen line is soft wrapped. However, I'm talking about the buffer line.

**Comment by ungb** [comment]
Also, what OS are you on?  I tried this on mac.

**Comment by Aerijo** [comment]
macOS. As the information in the initial post shows.

**Comment by ungb** [comment]
thanks @Aerijo, sorry I missed that detail.

I was actually able to repro the issue with a larger buffer line.  

Repro steps:
1. download https://code.jquery.com/jquery-3.3.1.min.js 
2. Open the file on Atom.
3. At the end of the buffer start typing

Expected: No lag while typing
Actual: Lags while typing

NOTE: if you hit enter and create a new line there is no lag.

**Comment by Aerijo** [comment]
> NOTE: if you hit enter and create a new line there is no lag when typing there.

@ungb Yes, the cause of the lag is because `autocomplete-plus` is trying to find the current word. To do this, it is applying a regex to the current line. When this line is extremely long, the regex execution takes a while to finish. This time taken translates to noticeable lag. There is a setting to increase the time between typing a character and the display popping up, but 
1) That doesn't remove the lag, just hides it, and 
2) I don't want a delay.

As you said, going to a new buffer line makes it normal again. This is because it's only searching for the word on the current buffer line, so it's completely ignoring the long line above it.

BTW, just to confirm you are reproducing the same issue, could you try again with `autocomplete-plus` disabled? I find the speed returns to normal when I do this. Additionally, making a performance profile should show it's the `exec` function taking up the time when the lag is occuring.

**Comment by maxbrunsfeld** [comment]
Thanks for the report @Aerijo! I'm not sure if this regressed recently or what. Seems like an easy fix though; we'll add it to our queue. Or, if you're interested in submitting a PR that limits the possible length of the prefix (say 80 characters or so?) I'd definitely merge it.
